mlflow/utils/req_parsing.py

```diff
@@ -0,0 +1,77 @@
+import os
+from itertools import filterfalse
+
+
+def _is_comment(line):
+    return line.startswith("#")
+
+
+def _is_empty(line):
+    return line == ""
+
+
+def _strip_inline_comment(line):
+    return line[: line.find(" #")].rstrip() if " #" in line else line
+
+
+def _is_requirements_file(line):
+    return line.startswith("-r ") or line.startswith("--requirement ")
+
+
+def _join_continued_lines(lines):
+    """
+    Joins lines ending with '\\'.
+
+    >>> _join_continued_lines["a\\", "b\\", "c"]
+    >>> 'abc'
+    """
+    continued_lines = []
+
+    for line in lines:
+        if line.endswith("\\"):
+            continued_lines.append(line.rstrip("\\"))
+        else:
+            continued_lines.append(line)
+            yield "".join(continued_lines)
+            continued_lines.clear()
+
+    # The last line ends with '\'
+    if continued_lines:
+        yield "".join(continued_lines)
+
+
+def _parse_requirements(requirements_file):
+    """
+    A simplified version of `pip._internal.req.parse_requirements` which performs the following
+    operations on the given requirements file and yields the parsed requirements.
+
+    - Remove comments and blank lines
+    - Join continued lines
+    - Resolve requirements file references (e.g. '-r requirements.txt')
+
+    References:
+    - `pip._internal.req.parse_requirements`:
+      https://github.com/pypa/pip/blob/7a77484a492c8f1e1f5ef24eaf71a43df9ea47eb/src/pip/_internal/req/req_file.py#L118
+    - Requirements File Format:
+      https://pip.pypa.io/en/stable/cli/pip_install/#requirements-file-format
```

Comment:
I noticed that constraint files are also supported by the requirements file format (`-c`). Do we need to add those?

Reply:
Thanks for the catch, yeah I think we do.

```
# a.txt
scikit-learn
-c b.txt
```

```
# b.txt
xgboost < 1.5
```

```
>>> pip install -r a.txt
Requirement already satisfied: scikit-learn in /Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/mlflow-dev-env/lib/python3.7/site-packages (from -r b.txt (line 1)) (0.24.1)
Requirement already satisfied: threadpoolctl>=2.0.0 in /Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/mlflow-dev-env/lib/python3.7/site-packages (from scikit-learn->-r b.txt (line 1)) (2.1.0)
Requirement already satisfied: scipy>=0.19.1 in /Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/mlflow-dev-env/lib/python3.7/site-packages (from scikit-learn->-r b.txt (line 1)) (1.6.2)
Requirement already satisfied: joblib>=0.11 in /Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/mlflow-dev-env/lib/python3.7/site-packages (from scikit-learn->-r b.txt (line 1)) (1.0.1)
Requirement already satisfied: numpy>=1.13.3 in /Users/harutakakawamura/.pyenv/versions/miniconda3-4.7.12/envs/mlflow-dev-env/lib/python3.7/site-packages (from scikit-learn->-r b.txt (line 1)) (1.21.0)
```

We could merge `a.txt` and `b.txt` into one list like:

```
scikit-learn
xgboost < 1.5
```

but this is actually incorrect.

Reply:
One thing we could do is collect constraint requirements in a single constraint file and bundle it with model.

Reply:
We can address this in a follow-up because this PR is getting larger.

Reply:
Offline discussion: we'll address this as a follow-up.

I'll add a TODO comment for this.